src/librustc_typeck/check/mod.rs

```diff
@@ -1228,13 +1233,25 @@ impl<'a, 'tcx> Visitor<'tcx> for GatherLocalsVisitor<'a, 'tcx> {
         intravisit::walk_local(self, local);
     }
 
+    fn visit_param(&mut self, param: &'tcx hir::Param<'tcx>) {
+        self.within_fn_param = true;
+        intravisit::walk_param(self, param);
+        self.within_fn_param = false;
+    }
+
     // Add pattern bindings.
     fn visit_pat(&mut self, p: &'tcx hir::Pat<'tcx>) {
         if let PatKind::Binding(_, _, ident, _) = p.kind {
             let var_ty = self.assign(p.span, p.hir_id, None);
 
-            if !self.fcx.tcx.features().unsized_locals {
-                self.fcx.require_type_is_sized(var_ty, p.span, traits::VariableType(p.hir_id));
+            if self.within_fn_param {
+                if !self.fcx.tcx.features().unsized_fn_params {
+                    self.fcx.require_type_is_sized(var_ty, p.span, traits::SizedArgumentType(None));
+                }
+            } else {
+                if !self.fcx.tcx.features().unsized_locals {
+                    self.fcx.require_type_is_sized(var_ty, p.span, traits::VariableType(p.hir_id));
+                }
             }
```

Comment:
What about `fn a(box box b: Box<Box<[u8]>>) {}`? I think that should keep using `unsized_locals` and not `unsized_fn_params`.

Reply:
Yes, I assume that pattern will generate MIR with unsized locals.

Reply:
Hmm yeah, confirmed that it won't work well in this PR. How can we catch such a case?

Reply:
I have no familiarity with HIR, but it must be possible to somehow check if the function argument is a "trivial" pattern, right?

Reply:
The [`params`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/struct.Body.html#structfield.params) should all have a `pat` that's immediately [`Binding`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.PatKind.html#variant.Binding).

Reply:
> I have no familiarity with HIR, but it must be possible to somehow check if the function argument is a "trivial" pattern, right?

I think `p.simple_ident` could be used here as well as https://github.com/rust-lang/rust/pull/74971/files#diff-1d1b0d29a2e8da97c6bfb6e364d920c7R1378-R1381?

Reply:
Maybe?^^ I'm really the wrong person to ask such questions.

Reply:
Hmm, anyway thanks for pointing out :) Let's move this topic to Zulip so that the we can get some eyes from other members...

Reply:
(summarizing my discussion with @spastorino on Zulip)
I think a bigger issue than special-casing the shape of the pattern, is that `within_fn_param` applies recursively to every nested pattern in a `fn` parameter, *instead of* to the parameter pattern itself.

So my suggestion is to replace it with a "shallow" version (bikeshed name: `outermost_fn_param_pat`) that `visit_pat` temporarily sets back to `false` around its `walk_pat` call, so that the `#![feature(unsized_fn_params)]` path is only taken for `fn` parameters themselves, *not* any other patterns nested in them.